src/rov/rov_pca9685/src/PCA9685.cpp

```diff
@@ -0,0 +1,190 @@
+    //     setAllOff();
+    // }
+    // for (uint8_t idx = 14; idx < 16; idx--) {
+    //     setOn(idx);
+    //     usleep(5000); 
+    //     setAllOff();
+    // }
+
+    // clear all registers
+    Reset();
+
+    return true;
+}
+
+void PCA9685::Close() {
+    setAllOff();
+    usleep(500);
+    if (i2c_fd > 0) {
+        close(i2c_fd);
+    }
+}
+
+void PCA9685::Reset() {
+    // stop all pwms
+    i2c_smbus_write_byte_data(i2c_fd, Registers::ALL_LED_OFF_H, BIT4);
+    usleep(500);
+    
+    uint8_t mode1_current = i2c_smbus_read_byte_data(i2c_fd, Registers::MODE1);
+    if (mode1_current & MODE1_RESTART) {
+        i2c_smbus_write_byte_data(i2c_fd, Registers::MODE1, mode1_current & ~MODE1_SLEEP);
+        usleep(500);
+    }
+    i2c_smbus_write_byte_data(i2c_fd, Registers::MODE1, (mode1_current & ~MODE1_SLEEP) | MODE1_RESTART);
+}
+
+void PCA9685::helloworld() {
+
+}
+
+void PCA9685::setFrequency(uint16_t frequency) {
+    static const uint16_t NUM_COUNTS = 4096; // scaling factor since clock sucks balls lol
+    // if given frequency is not valid, do nothing
+    if (frequency < 24 || frequency > 1526) {
+        return;
+    }
+    output_frequency = frequency;
+    uint8_t prescale = static_cast<uint8_t>(std::round(25e6/(NUM_COUNTS * frequency))) - 1;
+    // printf("prescale: 0x%02x", prescale);
+    uint8_t current_mode = i2c_smbus_read_byte_data(i2c_fd, Registers::MODE1);
+    // set sleep bit
+    i2c_smbus_write_byte_data(i2c_fd, Registers::MODE1, current_mode | MODE1_SLEEP);
+    usleep(500);
+    // set the prescalar
+    i2c_smbus_write_byte_data(i2c_fd, Registers::PRE_SCALE, prescale);
+    // reset the mode
+    i2c_smbus_write_byte_data(i2c_fd, Registers::MODE1, current_mode);
+
+    // restart
+    usleep(500);
+    i2c_smbus_write_byte_data(i2c_fd, Registers::MODE1, current_mode | MODE1_RESTART);
+}
+
+void PCA9685::setAllPWM(uint16_t on_counts, uint16_t off_counts) {
+    i2c_smbus_write_byte_data(i2c_fd, Registers::ALL_LED_ON_L, on_counts & 0xFF);
+    i2c_smbus_write_byte_data(i2c_fd, Registers::ALL_LED_ON_H, on_counts >> 8);
+    i2c_smbus_write_byte_data(i2c_fd, Registers::ALL_LED_OFF_L, off_counts & 0xFF);
+    i2c_smbus_write_byte_data(i2c_fd, Registers::ALL_LED_OFF_H, off_counts >> 8);
+}
+
+void PCA9685::setAllOff() {
+    // clear existing full on bit after setting full off (off takes precedence)
+    i2c_smbus_write_byte_data(i2c_fd, ALL_LED_OFF_H, BIT4);
+    i2c_smbus_write_byte_data(i2c_fd, ALL_LED_ON_H, i2c_smbus_read_byte_data(i2c_fd, ALL_LED_ON_H) & ~BIT4);
+}
+
+void PCA9685::setAllOn() {
+    // clear existing full off bit after setting full on (off takes precedence)
+    i2c_smbus_write_byte_data(i2c_fd, ALL_LED_ON_H, BIT4);
+    i2c_smbus_write_byte_data(i2c_fd, ALL_LED_OFF_H, i2c_smbus_read_byte_data(i2c_fd, ALL_LED_OFF_H) & ~BIT4);
+}
+
+void PCA9685::setDuty(uint8_t channel, float duty, float _delay) {
+    // uint16_t time_on = static_cast<uint16_t>(std::round(duty * 4096));
```

Comment:
yeet the comments

Reply:
comments are the "intended" way of doing this. could not get it to work with 0% duty cycle though since the math for the time_off register overflows at this condition...

This is actually a really hacky way of using the 9685. When counts <= time_off, the output is inverted. When only time_off is utilized, its actually the total number of counts that the pwm signal is asserted (leading edge at 0 counts, trailing edge at time_off counts).

If you could figure out how the math is supposed to work at this condition we would have a more functional 9685 driver (capable of offsetting the pulses wrt each other)

Reply:
@yameatmeyourdead thoughts on new impl